src/common/ui/code_extras/mode/javascript-mixed/javascript-mixed.js

```diff
@@ -0,0 +1,304 @@
+        // string template to obtain tokenQuasi tokenizer
+        const dummyStream = new stream.constructor('`#dummy', 2, {});
+        const dummyState = jsMode.startState();
+        jsMode.token(dummyStream, dummyState);
+        tokenQuasi = dummyState.tokenize;
+        return tokenQuasi;
+      }
+
+      function _forceJsModeToQuasi(stream, jsState) {
+        jsState.tokenize = getTokenQuasi(stream);
+      }
+
+      return _forceJsModeToQuasi;
+    }());
+
+
+    function prepareReparseStringTemplateInLocalMode(stream, state) {
+      dbg('Entering local hml/css mode...');
+      // spit out beginning backtick as a token, and leave the rest of the text for local mode parsing
+      stream.backUp(stream.current().length - 1);
+
+      // workaround needed for 1-line string template,
+      // to ensure the ending backtick is parsed correctly.
+      forceJsModeToQuasi(stream, state.jsState);
+    }
+
+    function exitLocalModeAndTokenEndingBacktick(stream, state) {
+      dbg('Exiting local html/css mode...');
+      // parse the ending JS string template backtick in js mode
+      return jsMode.token(stream, state.jsState);
+    }
+
+    function tokenInLocalMode(modeToUse, stream, state) {
+      // parse the text in local mode
+      state.localMode = state.localMode || modeToUse;
+      state.localState = state.localState || CodeMirror.startState(state.localMode);
+      const style = state.localMode.token(stream, state.localState);
+      dbg(`  local ${modeToUse.name} mode token - `, stream.current());
+      return style;
+    }
+
+    function matchRule(rules, stream, state) {
+      for (const r of rules) {
+        if (r.curContext === (state.maybeLocalContext || '<start>')) {
+          // dbg('  rule:', r.curContext, r.matchFn.toString());
+          const matched = r.run(stream, state);
+          if (matched) { break; }
+        }
+      }
+    }
+
+    class Rule {
+      constructor(curContext, matchFn, nextContext, caseMatchedFn, caseNotMatchedFn) {
+        this.curContext = curContext;
+        this.matchFn = matchFn;
+        this.nextContext = nextContext;
+        this.caseMatchedFn = caseMatchedFn; // optional
+        this.caseNotMatchedFn = caseNotMatchedFn; // optional
+      }
+
+      run(stream, state) {
+        if (this.matchFn()) {
+          state.maybeLocalContext = this.nextContext;
+          if (state.maybeLocalContext == null) {
+            // local mode done, reset
+            state.localMode = null;
+            state.localState = null;
+          }
+          if (this.caseMatchedFn) { this.caseMatchedFn(); }
+          return true;
+        } // case rule transition criteria not matched
+        if (this.caseNotMatchedFn) {
+          this.caseNotMatchedFn();
+        } else { // default not matched logic: reset local mode matching
+          state.maybeLocalContext = null;
+        }
+        return false;
+      }
+    }
+
+    const cssMatcher = (function () {
```

Comment:
I don't see the *physical* need for this encapsulation so I'd say `maybeToken` should be just `maybeCssToken` in the upper scope. I understand you copied the way it was structured in CodeMirror but does it warrant an extra scope?

Reply:
assume we move the rules outside the function, I'd think the encapsulation would still serve some purpose (safety, but also readability), grouping things needed by css matching only. E.g.,

```javascript
const cssMatcher = (function() {
  const cssMode = ... ;
  const cssRules = ...; // assuming we move it outside maybeToken()

  function maybeToken() { ... } // that uses cssMode/rules
  return { maybeToken };
})();
```

Reply:
Well, keep it if you like it but use ES6 `(() => {`.

Anyway, what I dislike in this setup is that the function is declared separately and is exported in an object so when jumping to definition or quickly looking up its code at a call site the extra step becomes an unnecessary pain. It's also unclear to me why use the extra complexity if there's only one function - why not export the function itself without all these indirections?

Reply:
> if there's only one function - why not export the function itself without all these indirections?
By that do you mean move the codes to a separate file and import the top level functions?

Anyway, I've removed the scope. The exposed structures are rather small in the end anyway.
```javascript
const cssMode
const cssRules
function maybeCssToken
// html counter part
```